### PR TITLE
coord,persist: teach coord about persisted streams that are involved in sources

### PR DESCRIPTION
**Pull request by aljoscha**
This is a refined take on #9403.
coord,persist: teach coord about persisted streams that are involved in sources

Before, the coordinator would only know the prefix that was used to form
the names of persistent streams that are involved in, say, the
persistent Kafka source.

Now, we keep structured information that is the same for all sources
(`primary_stream` and `timstamp_bindings_stream`) as well as
per-envelope persistence details. This way, the coordinator knows all
persisted streams that are involved in a source.

This also changes what persist information is sent to dataflow workers.
Before, we would send a single persist name. Now, we send a new
`SourcePersistDesc`, which contains the same structured persistence
details that the coordinator knows about. In future commits, we will add
more information to this.

We need to do this to allow the coordinator to:
 - determine a common seal timestamp when restarting
 - negotiate/figure out if a requested as_of matches the compaction
   frontier/since of a persisted source
 - compact persisted streams when necessary

We will do these additional changes in follow-up commits.

### Tips for reviewer

The individual commits are meaningful. The changes look bigger than they are, `dataflow,sources: factor import_source() into render_local_source() and render_external_source()` is just moving code around.

I think `catalog,persist: make SerializedSourcePersistDetails more structured` is possibly contentious: I don't know how forward/backwards compatible serializing enums and structs is. I left this as a separate commit that is easy to remove.

I also added an `IngestDesc` that does nothing and can't be used right now. Not sure we want to leave that in, but I added some description/`TODO`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by aljoscha** [comment]
This doesn't take into account all the changes we discussed "offline" yet, but I wanted to throw this out early.

File: src/dataflow-types/src/types.rs [hunk 1]
```diff
@@ -508,7 +508,65 @@ pub struct SourceDesc {
     /// to the output of the source.
     pub operators: Option<LinearOperator>,
     pub bare_desc: RelationDesc,
-    pub persisted_name: Option<String>,
+    pub persist_desc: Option<SourcePersistDesc>,
```
Comment (danhhz) [inline src/dataflow-types/src/types.rs:511]: let's make this look more like the SourceConnectorConnector we discussed offline

Comment (aljoscha) [reply src/dataflow-types/src/types.rs:511]: :+1:

Do you have any suggestion that is not `SourceConnectorConnector`? I'm having nightmares from Enterprise Java ... `ConnectorBuilderFactoryFactory` :see_no_evil:

**Comment by aljoscha** [comment]
I added new commits to address comments, this should now be in line with what we discussed offline and here, I hope. :sweat_smile:  I also updated the PR description. PTAL.

**Comment by danhhz** [comment]
Looking at this today!

**Review by danhhz — COMMENTED** [review]
this has some line item comments from while I was poking around but ignore them for now, I think we still need more iteration on the approach. I've now looked at this again and for a good number of hours trying to figure out what I think should happen here. ... I'm still not sure. there are a few things that make all this tough.

- first is that I still don't feel like I understand enough about source rendering to be able to reason about which pieces of info will be needed in ingestd and which will be needed in dataflowd.
- second, I find it someone odd that SourceConnector is used all the way from the return value of plan_create_source through SourceDesc all the way later in dataflow. I think to truly clean this up, we might want to split this up so we can annotate the various parts with persist info along the way (maybe splitting out the SourceConnector::External variant into a new ExternalSourceConnector struct that's also returned from plan_create_source). otoh, I'm in no way sure that's the right thing to do because of the above point

while I was trying to get my bearings here, I threw together a POC of an alternate approach that is cleaner in some ways. it adds a new ::Persist variant, which is used now for tables and would also be used later for persisted sources in dataflowd (dataflowd would never get a ::Local variant and eventually that one could be removed). it stores the source persistence info on ::External, which sorta feels reasonable to me, and it could continue to be there for a "legacy" non-persisted source in an dataflowd world. no idea what gets sent to ingestd. if that's helpful, here it is: https://github.com/MaterializeInc/materialize/compare/main...danhhz:persist_descs?expand=1

I hate how long this is dragging on, but given how long it's taking me to ramp up on this code, it's probably worth spending time to get it right. I don't want to hold up progress any more than I have to though, so I'll throw something on your calendar for another design session tomorrow and see if we can shake this out. sorry again and thanks for bearing with me!

**Comment by aljoscha** [comment]
Closing in favour of #9596.

**Pull request closed without merging**